--- xen-3.4.orig/xen-3.4.3/tools/pygrub/src/GrubConf.py	2010-05-29 18:49:18.000000000 +0200
+++ xen-3.4/xen-3.4.3/tools/pygrub/src/GrubConf.py	2010-07-28 21:50:23.000000000 +0200
@@ -135,7 +135,8 @@
 
         if self.commands.has_key(com):
             if self.commands[com] is not None:
-                setattr(self, self.commands[com], arg.strip())
+                if arg.strip():
+                    setattr(self, self.commands[com], arg.strip())
             else:
                 logging.info("Ignored image directive %s" %(com,))
         else:
