Bug #18863: Skip empty commands
--- a/tools/pygrub/src/GrubConf.py
+++ b/tools/pygrub/src/GrubConf.py
@@ -138,7 +138,8 @@ class GrubImage(_GrubImage):
 
         if self.commands.has_key(com):
             if self.commands[com] is not None:
-                setattr(self, self.commands[com], arg.strip())
+                if arg.strip():
+                    setattr(self, self.commands[com], arg.strip())
             else:
                 logging.info("Ignored image directive %s" %(com,))
         else:
