Bug #18403: Fix quoting of Shell-RegExps.
--- a/tools/hotplug/Linux/init.d/xendomains	2010-05-17 13:40:41.000000000 +0200
+++ b/tools/hotplug/Linux/init.d/xendomains	2010-05-17 13:41:04.000000000 +0200
@@ -182,11 +182,11 @@
 
 parseln()
 {
-    if [[ "$1" =~ "\(domain" ]]; then
+    if [[ "$1" =~ "(domain" ]]; then
         name=;id=
-    else if [[ "$1" =~ "\(name" ]]; then
+    else if [[ "$1" =~ "(name" ]]; then
         name=$(echo $1 | sed -e 's/^.*(name \(.*\))$/\1/')
-    else if [[ "$1" =~ "\(domid" ]]; then
+    else if [[ "$1" =~ "(domid" ]]; then
         id=$(echo $1 | sed -e 's/^.*(domid \(.*\))$/\1/')
     fi; fi; fi
 
