Index: config.php
===================================================================
--- config.php	(Revision 804)
+++ config.php	(Arbeitskopie)
@@ -3,22 +3,22 @@
 # basePath should represent the path from the DOCUMENT_ROOT
 # NOTE: to have the permalinks/REST-ful resource urls work, 
 # you'll need to make similar changes to the rewrite rules
-$basePath = "/";
+$basePath = "/univention-management-console-frontend-doc/";
 
-//$_base_url = "http://" . $_SERVER["HTTP_HOST"] . $basePath;
-$_base_url = "";
-$_site_name = "The Dojo Toolkit";
-$defVersion = "1.4";
-$dataDir = dirname(__FILE__) . "/data/";
+$_base_url = "http://" . $_SERVER["HTTP_HOST"] . $basePath;
+$_site_name = "Univention Management Console Frontend API Documentation";
+$defVersion = "umc";
+$dataDir = "/usr/share/univention-management-console-frontend-doc/data/";
 $defPage = "";
 $default_theme = "dtk";
 $theme = "dtk";
 $filter_privates = true;
 $use_cache = true;
 $modules = array(
	"dojo"=>-1, 
	"dijit"=>-1, 
	"dojox"=>-1, 
	"djConfig"=>-1
+	"umc"=>-1
  );
 ?>
Index: index.php
===================================================================
--- index.php	(Revision 804)
+++ index.php	(Arbeitskopie)
@@ -75,7 +75,7 @@
 <link rel="stylesheet" href="<?php echo $basePath; ?>themes/<?php echo $th; ?>/<?php echo $th; ?>.css" type="text/css" media="all" />
 <?php } ?>
 		<script type="text/javascript">djConfig={isDebug:false};</script>
-		<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/dojo/1.5/dojo/dojo.xd.js"></script>
+		<script type="text/javascript" src="/univention-dojo/dojo/dojo.js"></script>
 		<!-- SyntaxHighlighter -->
 		<script type="text/javascript" src="<?php echo $basePath; ?>js/syntaxhighlighter/scripts/shCore.js"><</script>
 		<script type="text/javascript" src="<?php echo $basePath; ?>js/syntaxhighlighter/scripts/shBrushJScript.js"><</script>
Index: .htaccess
===================================================================
--- .htaccess	(Revision 804)
+++ .htaccess	(Arbeitskopie)
@@ -1,14 +1,7 @@
-# supress php errors
-php_flag display_startup_errors off
-php_flag display_errors on
-php_flag html_errors off
-
-Options -MultiViews
-
 RewriteEngine On
 
 #  Set the prefix for all matches
-RewriteBase   /
+RewriteBase   /univention-management-console-doc/
 
 # convert static/rest-like requests
 RewriteRule ^html/([0-9.]+)/(.*)\.html$ lib/item.php?v=$1&p=$2 [L,QSA]
