REF: https://issues.kolab.org/issue3237

---
diff --git a/horde-webmail/lib/SyncML/Command/Alert.php b/horde-webmail/lib/SyncML/Command/Alert.php
index 0b004f6..6f4aa5b 100644
--- a/horde-webmail/lib/SyncML/Command/Alert.php
+++ b/horde-webmail/lib/SyncML/Command/Alert.php
@@ -287,6 +287,17 @@ class SyncML_Command_Alert extends SyncML_Command {
                 'Creating SyncML_Sync object for database '
                 . $this->_targetLocURI .  '; sync type ' . $synctype,
                 __FILE__, __LINE__, PEAR_LOG_DEBUG);
+            /* Synchronize databases with the IMAP backend */
+            $databases = array ('kronolith', 'turba', 'nag', 'mnemo');
+            foreach($databases as $app) {
+                $GLOBALS['registry']->callByPackage($app, 'synchronize');
+            }
+            if ($serverAnchorNext == time ()) {
+                // make sure that the timestamps of the synchronized items in
+                // the database are different from serverAnchorNext
+                sleep (1);
+            }
+            $serverAnchorNext = $GLOBALS['backend']->getCurrentTimeStamp();
             $sync = new SyncML_Sync($synctype,
                                     $this->_targetLocURI,
                                     $this->_sourceLocURI,
