diff -r 0717ca718d88 include/ddk/ndis.h
--- a/include/ddk/ndis.h	Tue Jul 08 21:35:58 2008 -0700
+++ b/include/ddk/ndis.h	Thu Jul 17 23:09:31 2008 -0700
@@ -52,12 +52,16 @@
 #if defined(NDIS50_MINIPORT)
 #ifndef NDIS50
 #define NDIS50
+#define NDIS_MINIPORT_MAJOR_VERSION 5
+#define NDIS_MINIPORT_MINOR_VERSION 0
 #endif
 #endif /* NDIS50_MINIPORT */
 
 #if defined(NDIS51_MINIPORT)
 #ifndef NDIS51
 #define NDIS51
+#define NDIS_MINIPORT_MAJOR_VERSION 5
+#define NDIS_MINIPORT_MINOR_VERSION 1
 #endif
 #endif /* NDIS51_MINIPORT */
 
@@ -399,19 +403,19 @@
 
 typedef struct _NDIS_PACKET {
   NDIS_PACKET_PRIVATE  Private;
-  union {
-    struct {
+  _ANONYMOUS_UNION union {
+    _ANONYMOUS_STRUCT struct {
       UCHAR  MiniportReserved[2 * sizeof(PVOID)];
       UCHAR  WrapperReserved[2 * sizeof(PVOID)];
-    } s1;
-    struct {
+    } DUMMYSTRUCTNAME;
+    _ANONYMOUS_STRUCT struct {
       UCHAR  MiniportReservedEx[3 * sizeof(PVOID)];
       UCHAR  WrapperReservedEx[sizeof(PVOID)];
-    } s2;
-    struct {
+    } DUMMYSTRUCTNAME;
+    _ANONYMOUS_STRUCT struct {
       UCHAR  MacReserved[4 * sizeof(PVOID)];
-    } s3;
-  } u;
+    } DUMMYSTRUCTNAME;
+  } DUMMYUNIONNAME;
   ULONG_PTR  Reserved[2];
   UCHAR  ProtocolReserved[1];
 } NDIS_PACKET, *PNDIS_PACKET, **PPNDIS_PACKET;
@@ -642,6 +646,11 @@
 #define NDIS_DEVICE_DISABLE_WAKE_ON_MAGIC_PACKET  0x00000080
 #define NDIS_DEVICE_DISABLE_WAKE_ON_PATTERN_MATCH 0x00000100
 
+/* _NDIS_DEVICE_PNP_EVENT */
+typedef enum _NDIS_DEVICE_PNP_EVENT
+{
+  Unused
+} NDIS_DEVICE_PNP_EVENT;
 
 /* OID_GEN_NETWORK_LAYER_ADDRESSES */
 typedef struct _NETWORK_ADDRESS {
@@ -915,6 +924,33 @@
   } V4ESP;
 } NDIS_TASK_IPSEC, *PNDIS_TASK_IPSEC;
 
+typedef enum _NDIS_ENCAPSULATION {
+  UNSPECIFIED_Encapsulation,
+  NULL_Encapsulation,
+  IEEE_802_3_Encapsulation,
+  IEEE_802_5_Encapsulation,
+  LLC_SNAP_ROUTED_Encapsulation,
+  LLC_SNAP_BRIDGED_Encapsulation
+} NDIS_ENCAPSULATION;
+
+typedef struct _NDIS_ENCAPSULATION_FORMAT {
+  NDIS_ENCAPSULATION  Encapsulation;
+  struct {
+    ULONG  FixedHeaderSize : 1;
+    ULONG  Reserved : 31;
+  } Flags;
+  ULONG  EncapsulationHeaderSize;
+} NDIS_ENCAPSULATION_FORMAT, *PNDIS_ENCAPSULATION_FORMAT;
+
+typedef struct _NDIS_TASK_OFFLOAD_HEADER
+{
+  ULONG Version;
+  ULONG Size;
+  ULONG Reserved;
+  UCHAR OffsetFirstTask;
+  NDIS_ENCAPSULATION_FORMAT EncapsulationFormat;
+} NDIS_TASK_OFFLOAD_HEADER, *PNDIS_TASK_OFFLOAD_HEADER;
+
 typedef struct _NDIS_TASK_OFFLOAD {
   ULONG  Version;
   ULONG  Size;
@@ -926,24 +962,6 @@
 
 /* NDIS_TASK_OFFLOAD_HEADER.Version constants */
 #define NDIS_TASK_OFFLOAD_VERSION 1
-
-typedef enum _NDIS_ENCAPSULATION {
-  UNSPECIFIED_Encapsulation,
-  NULL_Encapsulation,
-  IEEE_802_3_Encapsulation,
-  IEEE_802_5_Encapsulation,
-  LLC_SNAP_ROUTED_Encapsulation,
-  LLC_SNAP_BRIDGED_Encapsulation
-} NDIS_ENCAPSULATION;
-
-typedef struct _NDIS_ENCAPSULATION_FORMAT {
-  NDIS_ENCAPSULATION  Encapsulation;
-  struct {
-    ULONG  FixedHeaderSize : 1;
-    ULONG  Reserved : 31;
-  } Flags;
-  ULONG  EncapsulationHeaderSize;
-} NDIS_ENCAPSULATION_FORMAT, *PNDIS_ENCAPSULATION_FORMAT;
 
 typedef struct _NDIS_TASK_TCP_IP_CHECKSUM {
   struct {
@@ -1542,6 +1560,8 @@
   /*OUT*/ PNDIS_HANDLE  PoolHandle,
   /*IN*/ UINT  NumberOfDescriptors,
   /*IN*/ UINT  ProtocolReservedLength);
+
+#define PROTOCOL_RESERVED_SIZE_IN_PACKET (4 * sizeof(PVOID))
 
 NDISAPI
 VOID
@@ -1756,6 +1776,46 @@
 
 /*
  * VOID
+ * NdisGetFirstBufferFromPacketSafe(
+ * IN PNDIS_PACKET  _Packet,
+ * OUT PNDIS_BUFFER  * _FirstBuffer,
+ * OUT PVOID  * _FirstBufferVA,
+ * OUT PUINT  _FirstBufferLength,
+ * OUT PUINT  _TotalBufferLength),
+ * IN MM_PAGE_PRIORITY _Priority)
+ */
+#define NdisGetFirstBufferFromPacketSafe(_Packet,         \
+                                     _FirstBuffer,        \
+                                     _FirstBufferVA,      \
+                                     _FirstBufferLength,  \
+                                     _TotalBufferLength,  \
+                                     _Priority)           \
+{                                                         \
+  PNDIS_BUFFER _Buffer;                                   \
+                                                          \
+  _Buffer         = (_Packet)->Private.Head;              \
+  *(_FirstBuffer) = _Buffer;                              \
+  if (_Buffer != NULL)                                    \
+    {                                                     \
+            *(_FirstBufferVA)     = MmGetSystemAddressForMdlSafe(_Buffer, _Priority);  \
+            *(_FirstBufferLength) = MmGetMdlByteCount(_Buffer);         \
+            _Buffer = _Buffer->Next;                                    \
+                  *(_TotalBufferLength) = *(_FirstBufferLength);              \
+                  while (_Buffer != NULL) {                                   \
+                    *(_TotalBufferLength) += MmGetMdlByteCount(_Buffer);      \
+                    _Buffer = _Buffer->Next;                                  \
+                  }                                                           \
+    }                             \
+  else                            \
+    {                             \
+      *(_FirstBufferVA) = 0;      \
+      *(_FirstBufferLength) = 0;  \
+      *(_TotalBufferLength) = 0;  \
+    } \
+}
+
+/*
+ * VOID
  * NdisQueryBuffer(
  * IN PNDIS_BUFFER  Buffer,
  * OUT PVOID  *VirtualAddress OPTIONAL,
@@ -2092,7 +2152,7 @@
  * NdisReinitializePacket(
  * IN OUT  PNDIS_PACKET  Packet);
  */
-#define NdisReinitializePacketCounts(Packet)    \
+#define NdisReinitializePacket(Packet)          \
 {                                               \
 	(Packet)->Private.Head = (PNDIS_BUFFER)NULL;  \
 	(Packet)->Private.ValidCounts = FALSE;        \
@@ -2967,6 +3027,11 @@
   /*IN*/ PUCHAR  PacketBuffer,
   /*IN*/ UINT  PacketSize);
 
+#define NdisMIndicateReceivePacket(_handle, _packets, _number) \
+{  \
+  (*((PNDIS_MINIPORT_BLOCK)(_handle))->PacketIndicateHandler)(_handle, _packets, _number); \
+}
+
 NDISAPI
 VOID
 DDKAPI
@@ -3155,6 +3220,13 @@
 DDKAPI
 NdisPacketPoolUsage(
   /*IN*/ NDIS_HANDLE  PoolHandle);
+
+NTOSAPI
+VOID
+DDKAPI
+NdisSetPacketPoolProtocolId(
+  /*IN*/ NDIS_HANDLE PacketPoolHandle,
+  /*IN*/ UINT ProtocolId);
 
 NDISAPI
 NDIS_STATUS
@@ -4517,6 +4589,29 @@
 
 /* Routines for NDIS miniport drivers */
 
+NTOSAPI
+VOID
+DDKAPI
+NdisMInitializeTimer(
+  /*IN*/ /*OUT*/ PNDIS_MINIPORT_TIMER Timer,
+  /*IN*/ NDIS_HANDLE MiniportAdapterHandle,
+  /*IN*/ PNDIS_TIMER_FUNCTION TimerFunction,
+  /*IN*/ PVOID FunctionContext);
+
+NTOSAPI
+VOID
+DDKAPI
+NdisMSetPeriodicTimer(
+  /*IN*/ PNDIS_MINIPORT_TIMER Timer,
+  /*IN*/ UINT MillisecondsPeriod);
+
+NTOSAPI
+VOID
+DDKAPI
+NdisMCancelTimer(
+  /*IN*/ PNDIS_MINIPORT_TIMER Timer,
+  /*OUT*/ PBOOLEAN TimerCancelled);
+
 NDISAPI
 VOID
 DDKAPI
diff -r 0717ca718d88 include/ddk/ntddndis.h
--- a/include/ddk/ntddndis.h	Tue Jul 08 21:35:58 2008 -0700
+++ b/include/ddk/ntddndis.h	Thu Jul 17 23:09:31 2008 -0700
@@ -152,6 +152,8 @@
 #define OID_802_3_XMIT_TIMES_CRS_LOST     0x01020206
 #define OID_802_3_XMIT_LATE_COLLISIONS    0x01020207
 
+#define OID_TCP_TASK_OFFLOAD              0xFC010201
+
 /* OID_GEN_MINIPORT_INFO constants */
 #define NDIS_MINIPORT_BUS_MASTER                      0x00000001
 #define NDIS_MINIPORT_WDM_DRIVER                      0x00000002
diff -r 0717ca718d88 include/ddk/scsi.h
--- a/include/ddk/scsi.h	Tue Jul 08 21:35:58 2008 -0700
+++ b/include/ddk/scsi.h	Thu Jul 17 23:09:31 2008 -0700
@@ -1272,6 +1272,13 @@
 #define IOCTL_SCSI_MINIPORT_NOT_QUORUM_CAPABLE  ((FILE_DEVICE_SCSI << 16) + 0x0520)
 #define IOCTL_SCSI_MINIPORT_NOT_CLUSTER_CAPABLE ((FILE_DEVICE_SCSI << 16) + 0x0521)
 
+/* READ_TOC formats */
+#define READ_TOC_FORMAT_TOC         0x00
+#define READ_TOC_FORMAT_SESSION     0x01
+#define READ_TOC_FORMAT_FULL_TOC    0x02
+#define READ_TOC_FORMAT_PMA         0x03
+#define READ_TOC_FORMAT_ATIP        0x04
+
 /* Read Capacity Data. Returned in Big Endian format */
 typedef struct _READ_CAPACITY_DATA {
   ULONG  LogicalBlockAddress;
diff -r 0717ca718d88 include/ddk/winddk.h
--- a/include/ddk/winddk.h	Tue Jul 08 21:35:58 2008 -0700
+++ b/include/ddk/winddk.h	Thu Jul 17 23:09:31 2008 -0700
@@ -3828,7 +3828,7 @@
  *   VOID)
  */
 #define KeGetCurrentProcessorNumber() \
-  ((ULONG)KeGetCurrentKPCR()->ProcessorNumber)
+  ((ULONG)KeGetCurrentKPCR()->Number)
 
 
 #if  __USE_NTOSKRNL__
@@ -3842,13 +3842,13 @@
 LONG
 DDKFASTAPI
 InterlockedIncrement(
-  /*IN*/ PLONG  VOLATILE  Addend);
+  /*IN*/ LONG VOLATILE *Addend);
 
 NTOSAPI
 LONG
 DDKFASTAPI
 InterlockedDecrement(
-  /*IN*/ PLONG  VOLATILE  Addend);
+  /*IN*/ LONG VOLATILE *Addend);
 
 NTOSAPI
 LONG
@@ -4895,6 +4895,10 @@
 RtlStringFromGUID( 
   /*IN*/ REFGUID  Guid, 
   /*OUT*/ PUNICODE_STRING  GuidString);
+
+#define RtlStringCbCopyA(dst, dst_len, src) strncpy(dst, src, dst_len)
+#define RtlStringCbPrintfA(args...) snprintf(args)
+#define RtlStringCbVPrintfA(args...) vsnprintf(args)
 
 NTOSAPI
 BOOLEAN
@@ -7350,6 +7354,11 @@
 #define KeFlushIoBuffers(_Mdl, _ReadOperation, _DmaOperation)
 
 NTOSAPI
+VOID
+DDKAPI
+KeFlushQueuedDpcs(VOID);
+
+NTOSAPI
 PRKTHREAD
 DDKAPI
 KeGetCurrentThread(
@@ -7448,6 +7457,8 @@
 DDKAPI
 KeLeaveCriticalRegion(
   VOID);
+
+#define KeMemoryBarrier() asm("mfence;")
 
 NTOSAPI
 NTSTATUS
@@ -7719,38 +7730,18 @@
   /*IN*/ BOOLEAN  Alertable,
   /*IN*/ PLARGE_INTEGER  Timeout  /*OPTIONAL*/);
 
-#if defined(_X86_)
-
-NTOSAPI
-VOID
-FASTCALL
-KfLowerIrql(
-  /*IN*/ KIRQL  NewIrql);
-
-NTOSAPI
-KIRQL
-FASTCALL
-KfRaiseIrql(
-  /*IN*/ KIRQL  NewIrql);
-
-#define KeLowerIrql(a) KfLowerIrql(a)
-#define KeRaiseIrql(a,b) *(b) = KfRaiseIrql(a)
-
-#else
+NTOSAPI
+VOID
+DDKAPI
+KeRaiseIrql(
+  /*IN*/ KIRQL new_irql,
+  /*OUT*/ PKIRQL old_irql);
 
 NTOSAPI
 VOID
 DDKAPI
 KeLowerIrql(
-  /*IN*/ KIRQL  NewIrql);
-
-NTOSAPI
-KIRQL
-DDKAPI
-KeRaiseIrql(
-  /*IN*/ KIRQL  NewIrql);
-
-#endif
+  /*IN*/ KIRQL irql);
 
 NTOSAPI
 KIRQL
diff -r 0717ca718d88 lib/ddk/ndis.def
--- a/lib/ddk/ndis.def	Tue Jul 08 21:35:58 2008 -0700
+++ b/lib/ddk/ndis.def	Thu Jul 17 23:09:31 2008 -0700
@@ -130,8 +130,8 @@
 ;NdisImmediateWritePortUlong
 ;NdisImmediateWritePortUshort
 NdisImmediateWriteSharedMemory@16
-;NdisInitAnsiString DATA
-;NdisInitUnicodeString DATA
+NdisInitAnsiString@8
+NdisInitUnicodeString@8
 ;NdisInitializeEvent
 NdisInitializeReadWriteLock@4
 ;NdisInitializeString
@@ -149,7 +149,7 @@
 NdisMAllocateMapRegisters@20
 NdisMAllocateSharedMemory@20
 NdisMAllocateSharedMemoryAsync@16
-;NdisMCancelTimer
+NdisMCancelTimer@8
 NdisMCloseLog@4
 NdisMCmActivateVc@8
 NdisMCmCreateVc@16
@@ -179,7 +179,7 @@
 ;NdisMIndicateStatus
 ;NdisMIndicateStatusComplete
 NdisMInitializeScatterGatherDma@12
-;NdisMInitializeTimer
+NdisMInitializeTimer@16
 NdisMMapIoSpace@20
 ;NdisMPciAssignResources
 NdisMPromoteMiniport@4
@@ -202,7 +202,7 @@
 NdisMSetAttributesEx@20
 ;NdisMSetInformationComplete
 NdisMSetMiniportSecondary@8
-;NdisMSetPeriodicTimer
+NdisMSetPeriodicTimer@8
 ;NdisMSetTimer
 NdisMSleep@4
 ;NdisMStartBufferPhysicalMapping
@@ -252,7 +252,7 @@
 NdisSendPackets@12
 ;NdisSetEvent
 ;NdisSetPacketCancelId
-;NdisSetPacketPoolProtocolId
+NdisSetPacketPoolProtocolId@8
 ;NdisSetPacketStatus
 ;NdisSetProtocolFilter
 ;NdisSetTimer
diff -r 0717ca718d88 lib/ddk/ntoskrnl.def
--- a/lib/ddk/ntoskrnl.def	Tue Jul 08 21:35:58 2008 -0700
+++ b/lib/ddk/ntoskrnl.def	Thu Jul 17 23:09:31 2008 -0700
@@ -506,6 +506,7 @@
 ;KeFindConfigurationEntry
 ;KeFindConfigurationNextEntry
 ;KeFlushEntireTb
+KeFlushQueuedDpcs@0
 KeGetCurrentThread@0
 KeGetPreviousMode@0
 KeGetRecommendedSharedDataAlignment@0
@@ -542,7 +543,7 @@
 ;KeIsExecutingDpc
 KeLeaveCriticalRegion@0
 ;KeLoaderBlock DATA
-;KeNumberProcessors DATA
+KeNumberProcessors DATA
 ;KeProfileInterrupt
 ;KeProfileInterruptWithSource
 KePulseEvent@12
diff -r 0717ca718d88 lib/ddk/scsiport.def
--- a/lib/ddk/scsiport.def	Tue Jul 08 21:35:58 2008 -0700
+++ b/lib/ddk/scsiport.def	Thu Jul 17 23:09:31 2008 -0700
@@ -8,7 +8,7 @@
 ScsiPortFlushDma@4
 ScsiPortFreeDeviceBase@8
 ScsiPortGetBusData@24
-ScsiPortGetDeviceBase@24
+ScsiPortGetDeviceBase@28
 ScsiPortGetLogicalUnit@16
 ScsiPortGetPhysicalAddress@16
 ScsiPortGetSrb@20
