? Kolab_Server-0.5.0.tgz
? Kolab_Server-0.5.0_p1.patch
Index: a/framework/Kolab_Server/lib/Horde/Kolab/Server/Object/user.php
===================================================================
RCS file: /repository/framework/Kolab_Server/lib/Horde/Kolab/Server/Object/Attic/user.php,v
retrieving revision 1.2.2.11
diff -u -B -r1.2.2.11 user.php
--- a/framework/Kolab_Server/lib/Horde/Kolab/Server/Object/user.php	25 Apr 2009 08:56:33 -0000	1.2.2.11
+++ b/framework/Kolab_Server/lib/Horde/Kolab/Server/Object/user.php	25 Apr 2009 12:34:45 -0000
@@ -203,6 +203,17 @@
     }
 
     /**
+     * Get the group mail addresses for this object
+     *
+     * @return mixed|PEAR_Error An array of group addresses, false if no groups were
+     *                          found.
+     */
+    function getGroupAddresses()
+    {
+        return $this->_db->getGroupAddresses($this->_uid);
+    }
+
+    /**
      * Returns the server url of the given type for this user.
      *
      * This method is used to encapsulate multidomain support.
