diff --git a/horde-webmail/lib/Horde/Kolab/Storage/List.php b/horde-webmail/lib/Horde/Kolab/Storage/List.php
index a9bff36..a31ef82 100644
--- a/a/a/lib/Horde/Kolab/Storage/List.php
+++ b/a/a/lib/Horde/Kolab/Storage/List.php
@@ -136,7 +136,7 @@ class Kolab_List {
      */
     function &listFolders()
     {
-        if (!isset($this->_list)) {
+        if (!isset($this->_list) || is_a($this->_list, 'PEAR_Error')) {
             $session = &Horde_Kolab_Session::singleton();
             $imap = &$session->getImap();
             if (is_a($imap, 'PEAR_Error')) {
