--- core.schema.orig	2005-05-26 19:52:14.677384464 +0200
+++ core.schema	2005-05-26 19:54:20.044325792 +0200
@@ -387,8 +387,8 @@
 objectclass ( 2.5.6.6 NAME 'person'
 	DESC 'RFC2256: a person'
 	SUP top STRUCTURAL
-	MUST ( sn $ cn )
-	MAY ( userPassword $ telephoneNumber $ seeAlso $ description ) )
+	MUST cn
+	MAY ( sn $ userPassword $ telephoneNumber $ seeAlso $ description ) )
 
 objectclass ( 2.5.6.7 NAME 'organizationalPerson'
 	DESC 'RFC2256: an organizational person'
@@ -397,7 +397,7 @@
 		preferredDeliveryMethod $ telexNumber $ teletexTerminalIdentifier $
 		telephoneNumber $ internationaliSDNNumber $ 
 		facsimileTelephoneNumber $ street $ postOfficeBox $ postalCode $
-		postalAddress $ physicalDeliveryOfficeName $ ou $ st $ l ) )
+		postalAddress $ physicalDeliveryOfficeName $ ou $ st $ l $ organizationName ) )
 
 objectclass ( 2.5.6.8 NAME 'organizationalRole'
 	DESC 'RFC2256: an organizational role'
