commit 0482b557c761cdd96a640123bfb86200f667797c
Author: doughays-dojo <doughays@us.ibm.com>
Date:   Mon Aug 19 09:53:18 2013 -0700

    Fixes #17388.  Don't call preventDefault for touch events.

diff --git a/dijit/form/_ListBase.js b/dijit/form/_ListBase.js
index c06fdc1..b5b8f47 100644
--- a/dijit/form/_ListBase.js
+++ b/dijit/form/_ListBase.js
@@ -45,7 +45,9 @@ define([
 					eventType
 				),
 				function(evt){
-					evt.preventDefault();
+					if(!/^touch/.test(evt.type)){
+						evt.preventDefault();
+					}
 					self[callbackFuncName](evt, this);
 				}
 			));
