--- dojo-release-1.8.0-src/dojo/dom-style.js.orig	2012-09-04 18:33:32.000000000 +0200
+++ dojo-release-1.8.0-src/dojo/dom-style.js	2012-09-04 18:33:36.000000000 +0200
@@ -292,6 +292,9 @@
 		//	|	});
 
 		var n = dom.byId(node), l = arguments.length, op = (name == "opacity");
+		if (!n || !n.style) {
+			return;
+		}
 		name = _floatAliases[name] || name;
 		if(l == 3){
 			return op ? _setOpacity(n, value) : n.style[name] = value; // Number
