--- dojo-release-1.8.0-src/dojo/request/iframe.js.orig	2012-09-04 18:39:08.000000000 +0200
+++ dojo-release-1.8.0-src/dojo/request/iframe.js	2012-09-04 18:41:20.000000000 +0200
@@ -309,7 +309,11 @@
 						}
 					}else{
 						// 'json' and 'javascript' and 'text'
-						response.text = doc.getElementsByTagName('textarea')[0].value; // text
+						try {
+							response.text = doc.getElementsByTagName("textarea")[0].value; //text
+						} catch(err) {
+							response.text = doc.getElementsByTagName("body")[0].innerHTML; //text
+						}
 					}
 					handlers(response);
 				}else{
