--- dojo/dijit/Dialog.js.orig	2013-03-11 18:45:21.000000000 +0100
+++ dojo/dijit/Dialog.js	2013-03-11 18:45:24.000000000 +0100
@@ -83,8 +83,9 @@
 		// autofocus: Boolean
 		//		A Toggle to modify the default focus behavior of a Dialog, which
 		//		is to focus on the first dialog element after opening the dialog.
-		//		False will disable autofocusing. Default: true
-		autofocus: true,
+ 		//		False will disable autofocusing. Default: false (on mobile) and true
+ 		//		otherwise
+		autofocus: !has('touch'),
 
 		// _firstFocusItem: [private readonly] DomNode
 		//		The pointer to the first focusable node in the dialog.
