--- dojo-release-1.8.0-src/dojox/grid/enhanced/_FocusManager.js.orig	2012-07-21 10:59:18.000000000 +0200
+++ dojo-release-1.8.0-src/dojox/grid/enhanced/_FocusManager.js	2012-09-04 18:23:21.000000000 +0200
@@ -500,7 +500,7 @@
 		var didFocus = false;
 		this.inherited(arguments);
 		if(this._colHeadNode && html.style(this._colHeadNode, 'display') != "none"){
-			dijitFocus.focus(this._colHeadNode);
+			//dijitFocus.focus(this._colHeadNode);
 			this._stopEvent(evt);
 			didFocus = true;
 		}
@@ -549,7 +549,7 @@
 		//		Overwritten
 		this.inherited(arguments);
 		//EDG now will decorate event on header key events, if no focus, the cell will be wrong
-		dijitFocus.focus(this._colHeadNode);
+		//dijitFocus.focus(this._colHeadNode);
 	},
 	//---------------Content Area------------------------------------------
 	findAndFocusGridCell: function(){
