--- dojo-release-1.8.0-src/dojox/widget/Standby.js.orig	2012-06-13 23:31:22.000000000 +0200
+++ dojo-release-1.8.0-src/dojox/widget/Standby.js	2012-09-04 18:26:42.000000000 +0200
@@ -75,6 +75,10 @@
 	//		Default color for the translucent overlay is light gray.
 	color: "#C0C0C0",
 
+	// opacity: float
+	//		Controls the opacity for the overlay.
+	opacity: 0.75,
+
 	// duration: Integer
 	//		Integer defining how long the show and hide effects should take in milliseconds.
 	//		Defaults to 500
@@ -544,7 +548,7 @@
 			properties: {opacity: {start: 0, end: 1}},
 			onEnd: function(){
 				self.onShow();
-				delete self._anim;
+				//delete self._anim;
 			}
 		});
 		this._anim = fx.combine([underlayNodeAnim,imageAnim]);
