--- dijit/templates/MenuItem.html.orig	2020-11-02 12:59:23.123166601 +0100
+++ dijit/templates/MenuItem.html	2020-11-02 12:59:53.439173218 +0100
@@ -6,7 +6,7 @@
 		role="presentation"></td>
 	<td class="dijitReset dijitMenuItemAccelKey" style="display: none" data-dojo-attach-point="accelKeyNode"></td>
 	<td class="dijitReset dijitMenuArrowCell" role="presentation">
-		<span data-dojo-attach-point="arrowWrapper" style="visibility: hidden">
+		<span data-dojo-attach-point="arrowWrapper" class="dijitDisplayNone">
 			<span class="dijitInline dijitIcon dijitMenuExpand"></span>
 			<span class="dijitMenuExpandA11y">+</span>
 		</span>
--- dijit/PopupMenuItem.js.orig	2020-11-02 13:00:17.451179548 +0100
+++ dijit/PopupMenuItem.js	2020-11-02 13:01:07.791195741 +0100
@@ -1,13 +1,14 @@
 define([
 	"dojo/_base/declare", // declare
 	"dojo/dom-style", // domStyle.set
+	"dojo/dom-class",
 	"dojo/_base/lang",
 	"dojo/query", // query
 	"./popup",
 	"./registry",	// registry.byNode
 	"./MenuItem",
 	"./hccss"
-], function(declare, domStyle, lang, query, pm, registry, MenuItem){
+], function(declare, domStyle, domClass, lang, query, pm, registry, MenuItem){
 
 	// module:
 	//		dijit/PopupMenuItem
@@ -81,7 +82,7 @@
 
 			this.popup.domNode.style.display="none";
 			if(this.arrowWrapper){
-				domStyle.set(this.arrowWrapper, "visibility", "");
+				domClass.remove(this.arrowWrapper, "dijitDisplayNone");
 			}
 			this.focusNode.setAttribute("aria-haspopup", "true");
 		},
