--- dojo-release-1.8.0-src/dojox/mobile/build/build.sh.orig	2012-04-12 08:20:13.000000000 +0200
+++ dojo-release-1.8.0-src/dojox/mobile/build/build.sh	2012-09-04 18:17:44.000000000 +0200
@@ -14,10 +14,10 @@
 profile=mobile
 dir=release-mobile-separate
 #standalone=standaloneScrollable=true
-if [ "$1" == "single" ]; then
+if [ "$1" = "single" ]; then
   profile=mobile-all
 fi
-if [ "$1" == "single" ]; then
+if [ "$1" = "single" ]; then
   dir=release-mobile-single
 fi
 shift 1
--- dojo-release-1.8.0-src/util/shrinksafe/build.sh.orig	2012-09-05 10:17:43.000000000 +0200
+++ dojo-release-1.8.0-src/util/shrinksafe/build.sh	2012-09-05 10:17:54.000000000 +0200
@@ -15,7 +15,7 @@
 rm -rf bin
 
 # call build.sh test to run the unit tests immediately
-if [ "$TEST" == "test" ]; then
+if [ "$TEST" = "test" ]; then
 	echo "Running tests."
 	cd tests
 	./runner.sh  #| grep errors -1
