--- univention-dojo/dijit/ProgressBar.js.orig	2014-10-21 12:30:44.980000000 +0200
+++ univention-dojo/dijit/ProgressBar.js	2014-10-21 12:33:12.112000000 +0200
@@ -123,6 +123,8 @@
 
 			domClass.toggle(this.domNode, "dijitProgressBarIndeterminate", this.indeterminate);
 			domClass.toggle(this.domNode, "dijitProgressBarIndeterminateRtl", this.indeterminate && !this.isLeftToRight());
+			domClass.toggle(this.domNode, "dijitProgressBarGreaterThan50Percent", !this.indeterminate && percent >= 0.49);
+			domClass.toggle(this.domNode, "dijitProgressBarLowerThan50Percent", !this.indeterminate && percent < 0.49);
 
 			tip.style.width = (percent * 100) + "%";
 			this.onChange();
