--- dojo-release-1.8.0-src/dojo/dom-geometry.js.orig	2012-09-04 18:29:00.000000000 +0200
+++ dojo-release-1.8.0-src/dojo/dom-geometry.js	2012-09-04 18:29:17.000000000 +0200
@@ -525,6 +525,9 @@
 		// returns: Object
 
 		node = dom.byId(node);
+		if (node.nodeType != 1) {
+			return { x: 0, y: 0, w: 0, h: 0 };
+		}
 		var	db = win.body(node.ownerDocument),
 			ret = node.getBoundingClientRect();
 		ret = {x: ret.left, y: ret.top, w: ret.right - ret.left, h: ret.bottom - ret.top};
