diff --git libs/xstyle/css.js libs/xstyle/css.js
index 19df55c..8e354e8 100644
--- libs/xstyle/css.js
+++ libs/xstyle/css.js
@@ -37,7 +37,7 @@ define(["require"], function(moduleRequire){
 			}
 			function checkForParser(){
 				var parser = testElementStyle('x-parse', null, 'content');
-				if(parser && parser != 'none'){
+				if(parser && parser != 'none' && parser != 'normal'){
 					// TODO: wait for parser to load
 					require([eval(parser)], callback);
 				}else{
