diff -r fc20d46584e4 lib/Horde/iCalendar.php
--- a/lib/Horde/iCalendar.php	Fri Dec 05 17:01:35 2008 +0100
+++ b/lib/Horde/iCalendar.php	Fri Dec 05 17:03:16 2008 +0100
@@ -604,9 +604,12 @@
 
                 // Parse parameters.
                 if (!empty($parts[2])) {
-                    preg_match_all('/;(([^;=]*)(=([^;]*))?)/', $parts[2], $param_parts);
+                    preg_match_all('/;(([^;=]*)(=("[^"]*"|[^;]*))?)/', $parts[2], $param_parts);
                     foreach ($param_parts[2] as $key => $paramName) {
                         $paramValue = $param_parts[4][$key];
+                        if (preg_match('/"([^"]*)"/', $paramValue, $parts)) {
+                            $paramValue = $parts[1];
+                        }
                         $params[String::upper($paramName)] = $paramValue;
                     }
                 }
@@ -832,7 +835,30 @@
                     if ($param_value === null) {
                         $params_str .= ";$param_name";
                     } else {
-                        $params_str .= ";$param_name=$param_value";
+                        $len = strlen($param_value);
+                        $safe_value = '';
+                        $quote = false;
+                        for ($i = 0; $i < $len; ++$i) {
+                            $ord = ord($param_value[$i]);
+                            // Accept only valid characters.
+                            if ($ord == 9 || $ord == 32 || $ord == 33 ||
+                                ($ord >= 35 && $ord <= 126) ||
+                                $ord >= 128) {
+                                $safe_value .= $param_value[$i];
+                                /** 
+                                 * Characters above 128 do not need to be quoted
+                                 * as per RFC2445 but Outlook requires this.
+                                 */
+                                if ($ord == 44 || $ord == 58 || $ord == 59 ||
+                                    $ord >= 128) {
+                                    $quote = true;
+                                }
+                            }
+                        }
+                        if ($quote) {
+                            $safe_value = '"' . $safe_value . '"';
+                        }
+                        $params_str .= ";$param_name=$safe_value";
                     }
                 }
             }
