--- gollem/debian/gollem/usr/share/horde3/gollem/lib/Gollem.php.orig	2007-06-26 13:22:25.000000000 +0200
+++ gollem/debian/gollem/usr/share/horde3/gollem/lib/Gollem.php	2007-08-15 16:07:52.000000000 +0200
@@ -877,6 +877,9 @@
         if (!empty($GLOBALS['gollem_backends'][$backend]['hordeauth']) &&
             (strcasecmp($GLOBALS['gollem_backends'][$backend]['hordeauth'], 'full') === 0)) {
             return Auth::getAuth();
+        } else if (!empty($GLOBALS['gollem_backends'][$backend]['hordeauth']) &&
+            (strcasecmp($GLOBALS['gollem_backends'][$backend]['hordeauth'], 'mappedUser') === 0)) {
+            return Horde::callHook('_hook_default_username', array(Auth::getAuth()));
         } else {
             return Auth::getBareAuth();
         }
