diff --git a/.htaccess b/.htaccess
index 7beced1..b83f939 100644
--- a/.htaccess
+++ b/.htaccess
@@ -1,13 +1,7 @@
-php_flag display_startup_errors on
-php_flag display_errors on
-php_flag html_errors on
-
-Options -MultiViews
-
 RewriteEngine On
 
 #  Set the prefix for all matches
-RewriteBase   /api/
+RewriteBase   /univention-management-console-doc/
 
 # convert static/rest-like requests
 RewriteRule ^html/([0-9.]+)/(.*)\.html$ lib/item.php?v=$1&p=$2 [L,QSA]
diff --git a/config.php b/config.php
index 26ae0bf..7c3729f 100644
--- a/config.php
+++ b/config.php
@@ -3,36 +3,36 @@
 # basePath should represent the path from the DOCUMENT_ROOT
 # NOTE: to have the permalinks/REST-ful resource urls work, 
 # you'll need to make similar changes to the rewrite rules
-$basePath = "/api";
+$basePath = "/univention-management-console-doc";
 
 $_base_url = "//" . $_SERVER["HTTP_HOST"] . $basePath . "/";
 //$_base_url = "./";
 
 # Location of dojo, without trailing slash
-$dojoroot = "//ajax.googleapis.com/ajax/libs/dojo/1.8.0";
+$dojoroot = "/univention-dojo";
 # $dojoroot = "/trunk";	// local build
 
-$_site_name = "The Dojo Toolkit";
+$_site_name = "Univention Management Console Frontend API Documentation";
 
 # Initially display this version
 $defVersion = "1.8";
 
 # Location of data files for each version.   This directory should contain a subdirectory for each product version,
 # and the subdirectory should contain details.xml and tree.json
-$dataDir = dirname(__FILE__) . "/../api_data/";
+$dataDir = '/var/www/univention-management-console-doc/data/';
 
 # Bug database: if $bugdb is specified as a non-empty string, add an "error in the documentation?" link to the
 # bottom of each tab that will navigate to specified link appended with the tab's URL
-$bugdb = "https://docs.google.com/spreadsheet/viewform?hl=en_US&formkey=dFlDcHEyaHMwbEd4MFBObkNrX0E1MFE6MQ&entry_0=";
+$bugdb = "https://forge.univention.de/bugzilla/";
 
 # Path and URL to reference doc: if set, use the specified directory to check if corresponding reference doc exists,
 # and then use the url to generate a link to that reference doc.
 # The reference doc directory structure needs to be similar to api URLs, with version numbers for the top level
 # directories.
 $refdoc = array(
-	"dir" => dirname(__FILE__) . "/../reference-guide/",
-	"url" => "/reference-guide/",
-	"suffix" => ".html"
+	"dir" => '/var/www/univention-doc/',
+	"url" => "/univention-doc/",
+	"suffix" => ''
 );
 
 $defPage = "";
@@ -42,5 +42,5 @@ $filter_privates = true;
 
 # Set to false to regenerate the page for a module every time you view it.
 # Useful while working on documentation, or the doc parser and api viewers themselves.
-$use_cache = false;
+$use_cache = true;
 
