diff --git a/config.js b/config.js
index 62e9ee3..c0b67fc 100755
--- a/config.js
+++ b/config.js
@@ -28,11 +28,11 @@ define({
 	//    - id: The module ID of the exporter.
 	//    - config: Arbitrary exporter configuration object, passed as an argument to the exporter function.
 	exporters: [
-		{ id: 'dojov1', config: { details: 'details.xml', tree: 'tree.json' } }
+		{ id: 'dojov1', config: { details: '../details.xml', tree: '../tree.json' } }
 	],
 
 	show: {
-		warn:   false,
+		warn:   true,
 		info:   false,
 		debug:  false,
 		memory: true
@@ -41,8 +41,9 @@ define({
 	// Configuration data for module ID resolution and path remapping within the parser.
 	// TODO: Document
 	environmentConfig: {
-		basePath: '../trunk',
+		basePath: '../univention-dojo',
 		packages: {
+			umc: 'umc',
 			dojo: 'dojo',
 			dijit: 'dijit',
 			dojox: 'dojox',
diff --git a/parse.sh b/parse.sh
index fca624e..652c86e 100755
--- a/parse.sh
+++ b/parse.sh
@@ -1,2 +1,2 @@
 #!/bin/bash
-node dojo/dojo.js load=parse $@
+rhino dojo/dojo.js load=parse $@
