diff --git a/config.js b/config.js
index 62e9ee3..9e955b4 100755
--- a/config.js
+++ b/config.js
@@ -28,7 +28,7 @@ define({
 	//    - id: The module ID of the exporter.
 	//    - config: Arbitrary exporter configuration object, passed as an argument to the exporter function.
 	exporters: [
-		{ id: 'dojov1', config: { details: 'details.xml', tree: 'tree.json' } }
+		{ id: 'dojov1', config: { details: '../data/1.8/details.xml', tree: '../data/1.8/tree.json' } }
 	],
 
 	show: {
@@ -41,14 +41,18 @@ define({
 	// Configuration data for module ID resolution and path remapping within the parser.
 	// TODO: Document
 	environmentConfig: {
-		basePath: '../trunk',
+		basePath: '../univention-dojo',
 		packages: {
+			umc: 'umc',
 			dojo: 'dojo',
 			dijit: 'dijit',
 			dojox: 'dojox',
 			doh: 'util/doh'
 		},
 		excludePaths: [
+			/put-selector/,
+			/xstyle/,
+
 			// Non-API code
 			/\/(?:tests|nls|demos)\//,
 
diff --git a/lib/processor/dojodoc.js b/lib/processor/dojodoc.js
index 16277fe..f620376 100644
--- a/lib/processor/dojodoc.js
+++ b/lib/processor/dojodoc.js
@@ -218,7 +218,9 @@ define([
 			if (line.length && indentRe.exec(line)[0].length === keyIndent && line.length !== keyIndent) {
 				var keyLine = keyRe.exec(line);
 
-				key = standardKeys[keyLine[1]] || keyLine[1];
+				try {
+					key = standardKeys[keyLine[1]] || keyLine[1];
+				} catch(e) { continue; }
 
 				// Flag that when we read the next line, need to measure the indent
 				valueIndentRe = null;
@@ -496,4 +498,4 @@ define([
 			mixinMetadata(value.evaluated.metadata, metadata);
 		}
 	};
-});
\ No newline at end of file
+});
