Index: horde-webmailer/horde-webmail/kronolith/lib/Driver/kolab.php
===================================================================
--- horde-webmailer.orig/horde-webmail/kronolith/lib/Driver/kolab.php	2010-08-27 23:17:12.000000000 +0200
+++ horde-webmailer/horde-webmail/kronolith/lib/Driver/kolab.php	2010-08-28 00:49:41.000000000 +0200
@@ -1135,6 +1135,8 @@
      */
     function listEvents($startDate = null, $endDate = null, $hasAlarm = false)
     {
+        $this->_store->_data->synchronize();
+        $this->reset();
         $result = $this->synchronize();
         if (is_a($result, 'PEAR_Error')) {
             return $result;
Index: horde-webmailer/horde-webmail/mnemo/lib/Driver/kolab.php
===================================================================
--- horde-webmailer.orig/horde-webmail/mnemo/lib/Driver/kolab.php	2010-08-27 23:17:12.000000000 +0200
+++ horde-webmailer/horde-webmail/mnemo/lib/Driver/kolab.php	2010-08-28 00:47:40.000000000 +0200
@@ -719,6 +719,7 @@
      */
     function retrieve()
     {
+        $this->_store->_data->synchronize();
         $memos = array();
 
         $note_list = $this->_store->getObjects();
Index: horde-webmailer/horde-webmail/nag/lib/Driver/kolab.php
===================================================================
--- horde-webmailer.orig/horde-webmail/nag/lib/Driver/kolab.php	2010-08-27 23:17:12.000000000 +0200
+++ horde-webmailer/horde-webmail/nag/lib/Driver/kolab.php	2010-08-28 00:46:36.000000000 +0200
@@ -953,6 +953,7 @@
      */
     function retrieve($completed = 1)
     {
+        $this->_store->_data->synchronize();
         $dict = array();
         $tasks = new Nag_Task();
 
Index: horde-webmailer/horde-webmail/turba/lib/Driver/kolab.php
===================================================================
--- horde-webmailer.orig/horde-webmail/turba/lib/Driver/kolab.php	2010-08-27 23:17:12.000000000 +0200
+++ horde-webmailer/horde-webmail/turba/lib/Driver/kolab.php	2010-08-28 00:43:45.000000000 +0200
@@ -805,6 +805,7 @@
      */
     function _search($criteria, $fields)
     {
+        $this->_store->_data->synchronize();
         $result = $this->connect();
         if (is_a($result, 'PEAR_Error')) {
             return $result;
