From: Gunnar Wrobel <p@rdus.de>
Subject: [PATCH] t/kronolith/HK/GW/AdaptStartDateOfRecurrencesWhenEditing

REF: https://issues.kolab.org/issue3885

STATUS: MERGED

Signed-off-by: Gunnar Wrobel <p@rdus.de>


--- a/lib/Driver.php	2010-03-08 17:57:26.000000000 +0100
+++ b/lib/Driver.php	2010-03-08 17:58:13.000000000 +0100
@@ -1848,6 +1848,8 @@
         if ($recur !== null && $recur !== '') {
             if (!isset($this->recurrence)) {
                 $this->recurrence = new Horde_Date_Recurrence($this->start);
+            } else {
+                $this->recurrence->setRecurStart($this->start);
             }
             if (Util::getFormData('recur_enddate_type') == 'date') {
                 $recur_enddate = Util::getFormData('recur_enddate');
