From: Gunnar Wrobel <p@rdus.de>
Subject: [PATCH] t/kronolith/HK/GW/SendInvitationsByDefault

REF: https://issues.kolab.org/issue2966

STATUS: HACK

Signed-off-by: Gunnar Wrobel <p@rdus.de>

--- a/templates/edit/edit.inc	2010-05-16 11:32:56.122204796 +0200
+++ b/templates/edit/edit.inc	2010-05-16 11:33:19.730773324 +0200
@@ -373,7 +373,7 @@
 <tr>
  <td></td>
  <td colspan="4">
-  <input id="sendupdates" type="checkbox" class="checkbox" name="sendupdates" />
+  <input id="sendupdates" type="checkbox" class="checkbox" name="sendupdates" checked="checked" />
   <label for="sendupdates"><?php echo $event->isInitialized() ? _("Send updates to all attendees?") : _("Send invitations to all attendees?") ?></label>
  </td>
 </tr>
