--- a/win-pvdrivers.o/installer.wxs	2014-06-23 14:51:30.343833542 +0200
+++ b/win-pvdrivers/installer.wxs	2014-06-23 14:51:58.535832417 +0200
@@ -43,14 +43,6 @@
 		    Type="raw" />
     </Property>
 
-    <?ifdef env.CERT_PUBLIC_FILENAME?>
-    <?if $(env._BUILDARCH) = AMD64?>
-      <Condition Message="Testsigning must be on for x64 when installing on Vista or newer.">
-        <![CDATA[(VersionNT < 600) OR (SYSTEMSTARTOPTIONS >< "TESTSIGNING")]]>
-      </Condition>
-    <?endif ?>
-    <?endif ?>
-
     <Directory Id='TARGETDIR' Name='SourceDir'>
       <Directory Id='ProgramFilesFolder' Name='PFiles'>
         <Directory Id='XenProgramFilesDir' Name='Xen PV Drivers'>
--- a/win-pvdrivers.o/makedist.bat	2014-06-23 14:51:30.371833540 +0200
+++ b/win-pvdrivers/makedist.bat	2014-06-23 14:52:58.519830025 +0200
@@ -1,5 +1,5 @@
 @echo off
-IF NOT EXIST set_ddk_path.bat ECHO >set_ddk_path.bat SET DDK_PATH=C:\WinDDK\7600.16385.0
+IF NOT EXIST set_ddk_path.bat ECHO >set_ddk_path.bat SET DDK_PATH=C:\WinDDK\7600.16385.1
 IF NOT EXIST set_ddk_path_2k.bat ECHO >set_ddk_path_2k.bat SET DDK_PATH_2K=C:\WinDDK\6001.18002
 
 SET BUILD_NUMBER=1
@@ -7,7 +7,8 @@
 SET VERSION_MINOR=0
 for /f %%i in ('hg identify --num -r .') do set REVISION=%%i
 
-hg status | findstr "^M" && GOTO TEST_BUILD
+SET /A REVISION+=1
+rem hg status | findstr "^M" && GOTO TEST_BUILD
 SET GPLPV_VERSION=%VERSION_MAJOR%.%VERSION_MINOR%.%REVISION%
 GOTO DONE_VERSION
 
@@ -17,8 +18,8 @@
 
 :DONE_VERSION
 
-SET /A NEW_BUILD_NUMBER=%BUILD_NUMBER%+1
-ECHO >build_number.bat SET BUILD_NUMBER=%NEW_BUILD_NUMBER%
+rem SET /A NEW_BUILD_NUMBER=%BUILD_NUMBER%+1
+rem ECHO >build_number.bat SET BUILD_NUMBER=%NEW_BUILD_NUMBER%
 
 ECHO BUILDING %GPLPV_VERSION%
 
@@ -76,6 +77,6 @@
 IF EXIST BUILD*.WRN GOTO DONT_TAG
 IF EXIST BUILD*.ERR GOTO DONT_TAG
 
-hg tag %GPLPV_VERSION%
+rem hg tag %GPLPV_VERSION%
 
 :DONT_TAG
--- a/win-pvdrivers.o/wix.bat	2014-06-23 14:51:30.371833540 +0200
+++ b/win-pvdrivers/wix.bat	2014-06-23 14:53:16.851829294 +0200
@@ -14,7 +14,7 @@
 IF "%DDK_TARGET_OS%"=="WinNET" (SET MSIOS=2003)
 IF "%DDK_TARGET_OS%"=="WinLH" (SET MSIOS=Vista2008)
 
-SET MSINAME=gplpv_%MSIOS%%MSIARCH%_%GPLPV_VERSION%%MSIBUILD%.msi
+SET MSINAME=gplpv_%MSIOS%%MSIARCH%_signed_%GPLPV_VERSION%%MSIBUILD%.msi
 
 for /F %%x in ('DIR /B %BASEDIR%\redist\wdf\%_BUILDARCH%\WdfCoInstaller?????.dll') do set WDFFILENAME=%%x
 
