--- a/win-pvdrivers/installer.wxs
+++ b/win-pvdrivers/installer.wxs
@@ -43,14 +43,6 @@
 		    Type="raw" />
     </Property>
 
-    <?ifdef env.CERT_PUBLIC_FILENAME?>
-    <?if $(env._BUILDARCH) = AMD64?>
-      <Condition Message="Testsigning must be on for x64 when installing on Vista or newer.">
-        <![CDATA[(VersionNT < 600) OR (SYSTEMSTARTOPTIONS >< "TESTSIGNING")]]>
-      </Condition>
-    <?endif ?>
-    <?endif ?>
-
     <Directory Id='TARGETDIR' Name='SourceDir'>
       <Directory Id='ProgramFilesFolder' Name='PFiles'>
         <Directory Id='XenProgramFilesDir' Name='Xen PV Drivers'>
--- a/win-pvdrivers/makedist.bat
+++ b/win-pvdrivers/makedist.bat
@@ -1,5 +1,5 @@
 @echo off
-IF NOT EXIST set_ddk_path.bat ECHO >set_ddk_path.bat SET DDK_PATH=C:\WinDDK\7600.16385.0
+IF NOT EXIST set_ddk_path.bat ECHO >set_ddk_path.bat SET DDK_PATH=C:\WinDDK\7600.16385.1
 IF NOT EXIST set_ddk_path_2k.bat ECHO >set_ddk_path_2k.bat SET DDK_PATH_2K=C:\WinDDK\6001.18002
 
 SET VERSION=0.11.0
@@ -8,8 +8,8 @@ IF EXIST build_number.bat CALL build_num
 
 SET GPLPV_VERSION=%VERSION%.%BUILD_NUMBER%
 
-SET /A NEW_BUILD_NUMBER=%BUILD_NUMBER%+1
-ECHO >build_number.bat SET BUILD_NUMBER=%NEW_BUILD_NUMBER%
+rem SET /A NEW_BUILD_NUMBER=%BUILD_NUMBER%+1
+rem ECHO >build_number.bat SET BUILD_NUMBER=%NEW_BUILD_NUMBER%
 
 ECHO BUILDING %GPLPV_VERSION%
 
@@ -66,6 +66,6 @@ IF NOT EXIST TAG_HG GOTO DONT_TAG
 IF EXIST BUILD*.WRN GOTO DONT_TAG
 IF EXIST BUILD*.ERR GOTO DONT_TAG
 
-hg tag %GPLPV_VERSION%
+rem hg tag %GPLPV_VERSION%
 
 :DONT_TAG
\ No newline at end of file
--- a/win-pvdrivers/wix.bat
+++ b/win-pvdrivers/wix.bat
@@ -14,7 +14,7 @@ IF "%DDK_TARGET_OS%"=="WinXP" (SET MSIAR
 IF "%DDK_TARGET_OS%"=="WinNET" (SET MSIOS=2003)
 IF "%DDK_TARGET_OS%"=="WinLH" (SET MSIOS=Vista2008)
 
-SET MSINAME=gplpv_%MSIOS%%MSIARCH%_%GPLPV_VERSION%%MSIBUILD%.msi
+SET MSINAME=gplpv_%MSIOS%%MSIARCH%_signed_%GPLPV_VERSION%%MSIBUILD%.msi
 
 for /F %%x in ('DIR /B %BASEDIR%\redist\wdf\%_BUILDARCH%\WdfCoInstaller?????.dll') do set WDFFILENAME=%%x
 
