diff --git a/Makefile.am b/Makefile.am
index 15f4c2c..b400292 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,7 @@
 ACLOCAL_AMFLAGS=-I m4
 
 sasllibdir = @cyrus_sasl2_prefix@/lib/sasl2
-securitylibdir = $(DESTDIR)$(libdir)/security
+securitylibdir = $(libdir)/security
 
 sasllib_LTLIBRARIES = libsaml.la
 libsaml_la_SOURCES = cy2_saml.c saml.c plugin_common.c plugin_common.h saml.h
diff --git a/Makefile.in b/Makefile.in
index d968d4f..83df936 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -269,7 +269,7 @@ top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 ACLOCAL_AMFLAGS = -I m4
 sasllibdir = @cyrus_sasl2_prefix@/lib/sasl2
-securitylibdir = $(DESTDIR)$(libdir)/security
+securitylibdir = $(libdir)/security
 sasllib_LTLIBRARIES = libsaml.la
 libsaml_la_SOURCES = cy2_saml.c saml.c plugin_common.c plugin_common.h saml.h
 libsaml_la_LDFLAGS = -lsasl2 -llasso -module -version-info 2:0:2
