--- opt/scalix-tomcat/bin/sxtomcat-webapps.scalix 17:57:15.000000000 +0200
+++ opt/scalix-tomcat/bin/sxtomcat-webapps	2009-09-08 17:50:39.000000000 +0200
@@ -152,7 +152,30 @@
 			    rm -Rf "/etc/opt/scalix/$d"
 			elif [ -d "/opt/scalix/web/template/$d" ]
 			then
-				cp -pRf "/opt/scalix/web/template/$d" $(dirname $instdir)
+				# cp -pRf "/opt/scalix/web/template/$d" $(dirname $instdir)
+				# first copy the directory structure
+				for dd in $(find /opt/scalix/web/template/${d} -type d ); do
+					name=${dd#/opt/scalix/web/template/}
+					dir=$(dirname $instdir)
+					dest="${dir}/${name}"
+
+					if [ ! -x "${dest}" ]; then
+						mkdir -p ${dest}
+						chmod --reference=${dd} ${dest}
+					fi
+				done
+
+				for f in $(find /opt/scalix/web/template/$d -type f ); do
+					name=${f#/opt/scalix/web/template/}
+					dir=$(dirname $instdir)
+					dest="${dir}/${name}"
+					
+					if [ -e "${dest}" ]; then
+						cp -p $f ${dest}.dpkg-dist
+					else
+						cp -p $f ${dest}
+					fi
+				done
 			fi
             aw_count=$((aw_count + 1))
 		done
