--- sork-passwd/debian/sork-passwd/usr/share/horde3/passwd/scripts/kpasswd_expect.orig	2007-08-16 11:27:17.000000000 +0200
+++ sork-passwd/debian/sork-passwd/usr/share/horde3/passwd/scripts/kpasswd_expect	2007-08-16 11:28:12.000000000 +0200
@@ -13,7 +13,7 @@
 set pid [spawn /usr/bin/kpasswd $user]
 set timeout 1
 expect {
-    "Password for *: "
+    "s Password: "
     {
         send $oldpw\r
     }
@@ -25,7 +25,7 @@
 }
 
 expect {
-    "Enter new password: : "
+    "New password: "
     {
         send $newpw\r
     }
@@ -37,7 +37,7 @@
 }
 expect {
     eof {exit 1}
-    "Enter it again: : "
+    "Verifying - New password: "
     {
         send $newpw\r
     }
@@ -48,7 +48,7 @@
         close err
         exit 1
     }
-    "Password changed."
+    "Password changed"
     {
         close err
         exit 0
