Make /lib/security/ configurable for multi-arch
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,7 +1,7 @@
 ACLOCAL_AMFLAGS=-I m4
 
 sasllibdir = @cyrus_sasl2_prefix@/lib/sasl2
-securitylibdir = $(libdir)/security
+securitylibdir = @pamlibdir@
 
 sasllib_LTLIBRARIES = liboidc.la
 liboidc_la_SOURCES = cy2_oidc.c oidc.c plugin_common.c plugin_common.h oidc.h
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -31,6 +31,10 @@ AC_ARG_WITH(pam, [  --with-pam=prefix             Base installation of PAM],
 		 [CFLAGS="$CFLAGS -I${with_pam}/include -DWITH_PAM"; 
 		  LDFLAGS="$LDFLAGS -L${with_pam}/lib";
 		  pam_prefix="${with_pam}"])
+AC_ARG_WITH([pamlibdir],
+			AS_HELP_STRING([--with-pamlibdir=DIR], [directory for PAM modules]),
+			[],
+			[with_pamlibdir=/lib/security])
 AC_ARG_WITH(rhonabwy, [  --with-rhonabwy[=prefix]         Base installation of rhonabwy],
 		   [CFLAGS="$CFLAGS -I${with_rhonabwy}/include"; 
 		    LDFLAGS="$LDFLAGS -L${with_rhonabwy}/lib -lrhonabwy"])
@@ -51,5 +55,6 @@ AC_FUNC_MALLOC
 
 AC_SUBST([cyrus_sasl2_prefix])
 AC_SUBST([pam_prefix])
+AC_SUBST([pamlibdir], [$with_pamlibdir])
 AC_CONFIG_FILES([Makefile])
 AC_OUTPUT
