--- opt/scalix/install/OM-ADM.inst.scalix 17:56:57.000000000 +0200
+++ opt/scalix/install/OM-ADM.inst	2009-09-08 17:55:20.000000000 +0200
@@ -129,16 +129,23 @@
 		# The file in place is the same as the new version.
 		# Don't do the copy so we don't lose the OLD version if there is one.
 	else
+		##################################################################
+		# changed by Univention GmbH, because we don't want to overwrite
+		# the customer settings
+		#
 		if [ -f $OMDATA/sys/$FILE ]
 		then
-# MSG:  file has been replaced
-			omgetmsg -c mc_110 1 241 $OMDATA/sys/$FILE $OMDATA/sys/O$FILE
-			mv $OMDATA/sys/$FILE $OMDATA/sys/O$FILE
-		fi
-		cp $OMDATA/release.sys/$FILE $OMDATA/sys/$FILE
+			cp $OMDATA/release.sys/$FILE $OMDATA/sys/${FILE}.dpkg-dist
+			# MSG:  file has been replaced
+		#	omgetmsg -c mc_110 1 241 $OMDATA/sys/$FILE $OMDATA/sys/O$FILE
+		#	mv $OMDATA/sys/$FILE $OMDATA/sys/O$FILE
+		else
+			# the file doesn't exists
+			cp $OMDATA/release.sys/$FILE $OMDATA/sys/$FILE
 
-		chown scalix $OMDATA/sys/$FILE
-		chgrp scalix $OMDATA/sys/$FILE
+			chown scalix $OMDATA/sys/$FILE
+			chgrp scalix $OMDATA/sys/$FILE
+		fi
 	fi
 done
 
