--- a/xen-4.1.0/tools/libxc/xc_core.c     Sat May 28 09:08:21 2011 +0100
+++ b/xen-4.1.0/tools/libxc/xc_core.c     Sat May 28 09:09:40 2011 +0100
@@ -468,7 +468,6 @@

     int auto_translated_physmap;
     xen_pfn_t *p2m = NULL;
-    unsigned long p2m_size = 0;
     struct xen_dumpcore_p2m *p2m_array = NULL;

     uint64_t *pfn_array = NULL;
@@ -569,7 +568,7 @@
         }

         sts = xc_core_arch_map_p2m(xch, dinfo->guest_width, &info, live_shinfo,
-                                   &p2m, &p2m_size);
+                                   &p2m, &dinfo->p2m_size);
         if ( sts != 0 )
             goto out;
     }
