--- a/xen-4.1.0/tools/libxc/xc_domain_save.c      Tue May 24 08:19:39 2011 +0100
+++ b/xen-4.1.0/tools/libxc/xc_domain_save.c      Sat May 28 09:08:21 2011 +0100
@@ -861,7 +861,7 @@
  out:

     if ( !success && p2m )
-        munmap(p2m, P2M_FLL_ENTRIES * PAGE_SIZE);
+        munmap(p2m, P2M_FL_ENTRIES * PAGE_SIZE);

     if ( live_p2m_frame_list_list )
         munmap(live_p2m_frame_list_list, PAGE_SIZE);
@@ -1955,7 +1955,7 @@
         munmap(live_shinfo, PAGE_SIZE);

     if ( ctx->live_p2m )
-        munmap(ctx->live_p2m, P2M_FLL_ENTRIES * PAGE_SIZE);
+        munmap(ctx->live_p2m, P2M_FL_ENTRIES * PAGE_SIZE);

     if ( ctx->live_m2p )
         munmap(ctx->live_m2p, M2P_SIZE(ctx->max_mfn));
